buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/DistributionTest.kt

```diff
@@ -0,0 +1,148 @@
+        // the version currently differs between our dev infrastructure, so we only track the name and the architecture
+        current.getName() + " " + System.getProperty("os.arch")
+    }
+
+    @Internal
+    val gradleInstallationForTest: GradleInstallationForTestEnvironmentProvider = GradleInstallationForTestEnvironmentProvider(project)
+
+    @Internal
+    val binaryDistributions: BinaryDistributions = BinaryDistributions(project.layout)
+
+    @Internal
+    val libsRepository: LibsRepositoryEnvironmentProvider = LibsRepositoryEnvironmentProvider(project.layout)
+
+    init {
+        dependsOn(Callable { if (binaryDistributions.distributionsRequired) listOf("all", "bin", "src").map { ":distributions:${it}Zip" } else null })
+        dependsOn(Callable { if (binaryDistributions.binZipRequired) ":distributions:binZip" else null })
+        dependsOn(Callable { if (libsRepository.required) ":toolingApi:publishLocalArchives" else null })
+        jvmArgumentProviders.add(gradleInstallationForTest)
+        jvmArgumentProviders.add(BinaryDistributionsEnvironmentProvider(binaryDistributions))
+        jvmArgumentProviders.add(libsRepository)
+    }
+}
+
+fun <K, V> Map<K, V>.asSystemPropertyJvmArguments(): Iterable<String> = this.map { (key, value) -> "-D$key=$value" }
+
+class LibsRepositoryEnvironmentProvider(layout: ProjectLayout) : CommandLineArgumentProvider, Named {
+
+    @Internal
+    val dir: DirectoryProperty = layout.directoryProperty()
+
+    @Input
+    var required: Boolean = false
+
+    override fun asArguments(): Iterable<String> =
+        if (required) mapOf("integTest.libsRepo" to dir.asFile.get().absolutePath).asSystemPropertyJvmArguments() else emptyList()
+
+    override fun getName(): String = "libsRepository"
+}
+
+class GradleInstallationForTestEnvironmentProvider(project: Project) : CommandLineArgumentProvider, Named {
+
+    @Internal
+    val gradleHomeDir: DirectoryProperty = project.layout.directoryProperty()
+
+    @Internal
+    val gradleUserHomeDir: DirectoryProperty = project.layout.directoryProperty()
+
+    @Internal
+    val toolingApiShadedJarDir: DirectoryProperty = project.layout.directoryProperty()
+
+    /**
+     * The user home dir is not wiped out by clean.
+     * Move the daemon working space underneath the build dir so they don't pile up on CI.
+     */
+    @Internal
+    val daemonRegistry: DirectoryProperty = project.layout.directoryProperty()
+
+    @Nested
+    val gradleDistribution = GradleDistribution(project, gradleHomeDir)
+
+    override fun asArguments(): Iterable<String> =
+        mapOf(
+            "integTest.gradleHomeDir" to gradleHomeDir.asFile.get().absolutePath,
+            "integTest.gradleUserHomeDir" to gradleUserHomeDir.asFile.get().absolutePath,
+            "org.gradle.integtest.daemon.registry" to daemonRegistry.asFile.get().absolutePath,
+            "integTest.toolingApiShadedJarDir" to toolingApiShadedJarDir.getAsFile().get().absolutePath
```

Comment:
:x: can use `asFile` instead of `getAsFile()`

Reply:
Done.